# Ayiti Analytics capstone project

## Exploring Background of different professional in Haiti from LinkedIn

#### Data acquisition

##### Scrapping LinkedIn to find link profile of different people in Haiti

In [2]:
# Load libraries
import requests
import re
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
import time
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np

In [14]:
# Create values Selenium
USERNAME = "chosson.jeanfranco@gmail.com" 
PASSWORD =  "**********************"
linkedin = 'https://www.linkedin.com'

In [15]:
# Open Selenium
browser = webdriver.Chrome()
browser.get(linkedin)
time.sleep(3)
# Identify email and password inputs and enter in information
email = browser.find_element_by_name('session_key')
password = browser.find_element_by_name('session_password')
email.send_keys(USERNAME + Keys.RETURN)
password.send_keys(PASSWORD + Keys.RETURN)
time.sleep(3)

##  Step 1

I want to build a data set of at least one thousand profiles. 

LinkedIn puts restrictions on its page to prevent scrapping. 


When you use the LinkedIn filter to search, if you search for people and filter by a specific country, you can get 100 pages with less than 11 people per page. it can be 10 or 5 people per page. 
In order to build my dataset, i will scrape linkedIn by searching for people and filter one by one for Haiti, some universities in Haiti, some companies in Haiti.

In [ ]:
# Go to search page and scroll to the bottom of the page to load elements of the page
search = "https://www.linkedin.com/search/results/people/?facetGeoRegion=%5B%22ht%3A0%22%5D&origin=FACETED_SEARCH"
browser.get(search)
time.sleep(3)
browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
current_url = 'url_placeholder' # this is a placeholder for the URL check
# Create empty dataframe
df = pd.DataFrame(columns = ['name', 'title', 'location', 'profile'])

debut_url = 'https://www.linkedin.com/search/results/people/?keywords=haiti&origin=SWITCH_SEARCH_VERTICAL&page='

for i in range(100):
    Final_url = debut_url + str(i)
    browser.get(Final_url)
    time.sleep(3)
    
    page = BeautifulSoup(browser.page_source, 'lxml')
    page_names = page.find_all('span', class_ = 'actor-name')
    page_titles = page.find_all('p', class_ = 'subline-level-1')
    page_locations = page.find_all('p', class_ = 'subline-level-2')
    page_profiles = page.find_all('a', class_ = 'search-result__result-link')
    
    # Put scraped data into a dataframe
    names = list(map(lambda x: x.text, page_names))
    titles = list(map(lambda x: x.text.replace('\n', ''), page_titles))
    locations = list(map(lambda x: x.text.replace('\n', ''), page_locations))
    profiles = list(map(lambda x: linkedin + x['href'], page_profiles))[::2]
    temp = pd.DataFrame({'name':names, 'title':titles, 'location':locations, 'profile':profiles})
    
    # Filter out members who do not provide information
    temp = temp[temp['name'] != 'LinkedIn Member']
    if temp.shape[0]<10:
        print(len(temp))
    # Append new data to df
    df = df.append(temp)
    
     # Stop appending if the number of retrieved records exceeds the limit
    if df.shape[0] >= 1000:
        break
    time.sleep(3)
        
# Reset dataframe index
df.reset_index()

# Export results
#df.to_csv("output_search.csv", index = False)

# Close Selenium
browser.quit()

In [19]:
# Create values Selenium
USERNAME = "chosson.jeanfranco@gmail.com" 
PASSWORD =  "**********************"
linkedin = 'https://www.linkedin.com'
# Open Selenium
browser = webdriver.Chrome()
browser.get(linkedin)
time.sleep(3)
# Identify email and password inputs and enter in information
email = browser.find_element_by_name('session_key')
password = browser.find_element_by_name('session_password')
email.send_keys(USERNAME + Keys.RETURN)
password.send_keys(PASSWORD + Keys.RETURN)
# Go to profile page and scroll to the bottom of the page to load elements of the page
time.sleep(3)
search = "https://www.linkedin.com/search/results/people/?facetSchool=%5B%22162070%22%2C%225097000%22%5D&origin=FACETED_SEARCH"
browser.get(search)
time.sleep(3)
browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
current_url = 'url_placeholder' # this is a placeholder for the URL check
# Create empty dataframe
df = pd.DataFrame(columns = ['name', 'title', 'location', 'profile'])

debut_url = 'https://www.linkedin.com/search/results/people/?facetSchool=%5B%22162070%22%2C%225097000%22%5D&origin=FACETED_SEARCH&page='

for i in range(100):
    Final_url = debut_url + str(i)
    browser.get(Final_url)
    time.sleep(3)
    
    page = BeautifulSoup(browser.page_source, 'lxml')
    page_names = page.find_all('span', class_ = 'actor-name')
    page_titles = page.find_all('p', class_ = 'subline-level-1')
    page_locations = page.find_all('p', class_ = 'subline-level-2')
    page_profiles = page.find_all('a', class_ = 'search-result__result-link')
    
    # Put scraped data into a dataframe
    names = list(map(lambda x: x.text, page_names))
    titles = list(map(lambda x: x.text.replace('\n', ''), page_titles))
    locations = list(map(lambda x: x.text.replace('\n', ''), page_locations))
    profiles = list(map(lambda x: linkedin + x['href'], page_profiles))[::2]
    temp = pd.DataFrame({'name':names, 'title':titles, 'location':locations, 'profile':profiles})
    
    # Filter out members who do not provide information
    temp = temp[temp['name'] != 'LinkedIn Member']
    
    # Append new data to df
    df = df.append(temp)
    
     # Stop appending if the number of retrieved records exceeds the limit
    if df.shape[0] >= 1000:
        break
    time.sleep(3)
        
# Reset dataframe index
df.reset_index()

# Export results
df.to_csv("output_search1.csv", index = False)

# Close Selenium
browser.quit()

In [22]:
# Create values Selenium
USERNAME = "chosson.jeanfranco@gmail.com" 
PASSWORD =  "*************************"
linkedin = 'https://www.linkedin.com'
# Open Selenium
browser = webdriver.Chrome()
browser.get(linkedin)
time.sleep(3)
# Identify email and password inputs and enter in information
email = browser.find_element_by_name('session_key')
password = browser.find_element_by_name('session_password')
email.send_keys(USERNAME + Keys.RETURN)
password.send_keys(PASSWORD + Keys.RETURN)
# Go to profile page and scroll to the bottom of the page to load elements of the page
time.sleep(3)

search = "https://www.linkedin.com/search/results/people/?facetPastCompany=%5B%22163457%22%2C%22767299%22%2C%225654%22%2C%221860%22%2C%221964533%22%5D&origin=FACETED_SEARCH"
browser.get(search)
time.sleep(3)
browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
current_url = 'url_placeholder' # this is a placeholder for the URL check
# Create empty dataframe
df = pd.DataFrame(columns = ['name', 'title', 'location', 'profile'])

debut_url = 'https://www.linkedin.com/search/results/people/?facetPastCompany=%5B%22163457%22%2C%22767299%22%2C%225654%22%2C%221860%22%2C%221964533%22%5D&origin=FACETED_SEARCH&page='

for i in range(100):
    Final_url = debut_url + str(i)
    browser.get(Final_url)
    time.sleep(3)
    
    page = BeautifulSoup(browser.page_source, 'lxml')
    page_names = page.find_all('span', class_ = 'actor-name')
    page_titles = page.find_all('p', class_ = 'subline-level-1')
    page_locations = page.find_all('p', class_ = 'subline-level-2')
    page_profiles = page.find_all('a', class_ = 'search-result__result-link')
    
    # Put scraped data into a dataframe
    names = list(map(lambda x: x.text, page_names))
    titles = list(map(lambda x: x.text.replace('\n', ''), page_titles))
    locations = list(map(lambda x: x.text.replace('\n', ''), page_locations))
    profiles = list(map(lambda x: linkedin + x['href'], page_profiles))[::2]
    temp = pd.DataFrame({'name':names, 'title':titles, 'location':locations, 'profile':profiles})
    
    # Filter out members who do not provide information
    temp = temp[temp['name'] != 'LinkedIn Member']
    
    # Append new data to df
    df = df.append(temp)
    
     # Stop appending if the number of retrieved records exceeds the limit
    if df.shape[0] >= 1000:
        break
    time.sleep(3)
        
# Reset dataframe index
df.reset_index()

# Export results
df.to_csv("output_search2.csv", index = False)

# Close Selenium
browser.quit()

##### Load the three datasets

In [26]:
df = pd.read_csv('output_search.csv')
df1 = pd.read_csv('output_search1.csv')
df2 = pd.read_csv('output_search2.csv')

In [33]:
## Concat the three datasets into result
result = pd.concat([df, df1, df2], axis=0)

###### Drop duplicates values on profile

In [34]:
result.drop_duplicates(subset ="profile", keep = False, inplace = True)

In [44]:
### make sure all the profile is from people in Haiti
new_result = result[result['location'].str.contains('Haiti')]

In [46]:
## save dataset into a csv file.
new_result.to_csv("result.csv", index = False)

In [47]:
new_result['location'].value_counts()

        Haiti                                       1050
        Port-au-Prince, Haiti                         57
        Port-au-Prince Arrondissement, Haiti           8
        Pétionville, Haiti                             4
        Gros-Morne Arrondissement, Haiti               1
        Jacmel Arrondissement, Haiti                   1
        Trou-du-Nord Arrondissement, Haiti             1
        Carrefour, Haiti                               1
        Les Cayes Arrondissement, Haiti                1
        Delmas, Haiti                                  1
Name: location, dtype: int64

In [48]:
new_result.shape

(1125, 4)

##### We have dataset of 1125 profiles

## Step 2

From those profiles, i want to get their background, education, (Bachelor degree's or master degree's), their study fields, Their experiences and their skills. 
i'm gonna do another scrape, to get those informations from the differents profiles i get before.

In [ ]:
r = pd.read_csv("result.csv")

In [ ]:
#create empty data frame
Exp_df = pd.DataFrame(columns = ['profile', 'exp_title', 'exp_company', 'exp_dates'])
Edu_df = pd.DataFrame(columns = ['profile', 'ed_name', 'ed_deg', 'ed_dates'])
Ski_df = pd.DataFrame(columns = ['profile', 'skill'])

tcounter = 0 
#Create big loop
#for link in r.loc[0:5,'profile']:
for link in r.loc[:,'profile']:
    if link == 'https://www.linkedin.com#':     #if it equal link then skip
        continue 
    time.sleep(3)
# This section is where you put in the profile link (loaded from the csv file) and browse to it
    search = link
    browser.get(search)
    time.sleep(5)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    #browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    #raw = urlopen(link).read()
    #page = BeautifulSoup(raw, "html.parser")
    page = BeautifulSoup(browser.page_source, 'lxml')
    
    tcounter += 1
    
    if tcounter % 100 == 0:
        time.sleep(1800)
    
#Experience Section  
    titles = page.find_all('div', class_ = "pv-entity__position-group-pager")
    companies = page.find_all('span', class_ = "pv-entity__secondary-title")
    dates = page.find_all('h4', class_ = "pv-entity__date-range")

    #Put scraped data into exp_df

    arraylen1 = len(page.find_all('div', class_ = "pv-entity__position-group-pager"))

    profile = link
    exp_titles = list(map(lambda x: x.h3.text.strip(), titles))[0:arraylen1]
    exp_companies = list(map(lambda x: x.text.strip(), companies))[0:arraylen1]
    exp_dates = list(map(lambda x: x.text.strip().split('\n')[-1], dates))[0:arraylen1]
    
 #Education Section 
    institution = page.find_all('div', class_ = "pv-entity__degree-info")
    degree = page.find_all('p', class_ = "pv-entity__degree-name")
    dates = page.find_all('p', class_ = "pv-entity__dates")
 
    #Put scraped data into edu_df
    
    arraylen2 = len(page.find_all('div', class_ = "pv-entity__degree-info"))

    profile = link
    ed_name = list(map(lambda x: x.text.strip().split('\n')[-1], institution))[0:arraylen2]
    ed_deg = list(map(lambda x: x.text.strip().split('\n')[-1], degree))[0:arraylen2]
    ed_dates = list(map(lambda x: x.text.strip().split('\n')[-1], dates))[0:arraylen2]
    if len(ed_dates) < arraylen2:
        ed_dates = 'NA'
 #Skill Section 
    skill = page.find_all('span', class_ = "pv-skill-category-entity__name-text")
    
    #Put scraped data into a ski_df
    
    arraylen3 = len(page.find_all('span', class_ = "pv-skill-category-entity__name-text"))
        
    profile = link
    skill = list(map(lambda x: x.text.strip(), skill))[0:arraylen3]
    temp1 = pd.DataFrame({'profile':profile, 'exp_title':exp_titles, 'exp_company':exp_companies, 'exp_dates':exp_dates})
    temp2 = pd.DataFrame({'profile':profile, 'ed_name':ed_name, 'ed_deg':ed_deg, 'ed_dates':ed_dates}) 
    temp3 = pd.DataFrame({'profile':profile, 'skill':skill})
    Exp_df = Exp_df.append(temp1)
    Edu_df = Edu_df.append(temp2)
    Ski_df = Ski_df.append(temp3)

I've been dealing with linkedIn's restrictions such as captcha, and I'm being asked to re-authenticate after an amount of request. 
so i need to repeat the process several times. 

In [ ]:
#create empty data frame
Exp_df = pd.DataFrame(columns = ['profile', 'exp_title', 'exp_company', 'exp_dates'])
Edu_df = pd.DataFrame(columns = ['profile', 'ed_name', 'ed_deg', 'ed_dates'])
Ski_df = pd.DataFrame(columns = ['profile', 'skill'])

tcounter = 0 
#Create big loop
#for link in r.loc[0:5,'profile']:
for link in r.loc[:,'profile']:
    if link == 'https://www.linkedin.com#':     #if it equal link then skip
        continue 
    time.sleep(5)
# This section is where you put in the profile link (loaded from the csv file) and browse to it
    search = link
    browser.get(search)
    time.sleep(5)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    #browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    #raw = urlopen(link).read()
    #page = BeautifulSoup(raw, "html.parser")
    page = BeautifulSoup(browser.page_source, 'lxml')
    
    tcounter += 1
    
    if tcounter % 100 == 0:
        time.sleep(1000)
    
    
#Experience Section
    tall = page.find_all('li', class_ = "pv-entity__position-group-pager pv-profile-section__list-item ember-view")
    titles = []
    companies = []
    dates = []
    
    for li in tall:
        _title = li.find_all('h3','t-16 t-black t-bold')
        _companies = li.find_all('p','pv-entity__secondary-title t-14 t-black t-normal')
        _dates = li.find_all('span', class_ = "pv-entity__bullet-item-v2")
        if _title:
            titles.append(_title[0].get_text().strip()) 
        else:
             titles.append('')
        if _companies:
            companies.append(_companies[0].get_text().strip()) 
        else:
             companies.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
        print(companies)
    #Put scraped data into exp_df


    profile = link
    exp_titles = titles
    exp_companies =  companies
    exp_dates = dates
    
    #Education Section
    tall = page.find_all('li', class_ = "pv-profile-section__list-item pv-education-entity pv-profile-section__card-item ember-view")
    study_name = [] 
    degree = []
    dates = []
    
    for li in tall:
        _std_name = li.find_all('p','pv-entity__secondary-title pv-entity__fos t-14 t-black t-normal')
        _deg = li.find_all('p','pv-entity__secondary-title pv-entity__degree-name t-14 t-black t-normal')
        _dates = li.find_all('p', class_ = "pv-entity__dates t-14 t-black--light t-normal  time")
        if _std_name:
            _study_name = _std_name[0].find_all('span', 'pv-entity__comma-item')
            study_name.append(_study_name[0].get_text().strip()) 
        else:
             study_name.append('')
        if _deg:
            _degree = _deg[0].find_all('span', 'pv-entity__comma-item')
            degree.append(_degree[0].get_text().strip()) 
        else:
             degree.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
                
    ed_name = study_name
    ed_deg = degree
    ed_dates = dates
                
    #Skill Section 
    tall = page.find_all('li', class_ = "pv-skill-category-entity__top-skill pv-skill-category-entity pb3 pt4 pv-skill-endorsedSkill-entity relative ember-view")
    skill = []
    
    for li in tall:
        _skill = li.find_all('span','pv-skill-category-entity__name-text t-16 t-black t-bold')
        if _skill:
            skill.append(_skill[0].get_text().strip()) 
        else:
             skill.append('')

        
    temp1 = pd.DataFrame({'profile':profile, 'exp_title':exp_titles, 'exp_company':exp_companies, 'exp_dates':exp_dates})
    temp2 = pd.DataFrame({'profile':profile, 'ed_name':ed_name, 'ed_deg':ed_deg, 'ed_dates':ed_dates}) 
    temp3 = pd.DataFrame({'profile':profile, 'skill':skill})
    Exp_df = Exp_df.append(temp1)
    Edu_df = Edu_df.append(temp2)
    Ski_df = Ski_df.append(temp3)
    print(link, 'completed')
   

Exp_df.to_csv('experience.csv')
Edu_df.to_csv('education.csv')
Ski_df.to_csv('skill.csv')

### I was able to scrape 559 profiles

In [ ]:
newR = r.tail(566)

In [ ]:
#create empty data frame
newExp_df = pd.DataFrame(columns = ['profile', 'exp_title', 'exp_company', 'exp_dates'])
newEdu_df = pd.DataFrame(columns = ['profile', 'ed_name', 'ed_deg', 'ed_dates'])
newSki_df = pd.DataFrame(columns = ['profile', 'skill'])

tcounter = 0 
#Create big loop
#for link in r.loc[0:5,'profile']:
for link in newR.loc[:,'profile']:
    if link == 'https://www.linkedin.com#':     #if it equal link then skip
        continue 
    time.sleep(5)
# This section is where you put in the profile link (loaded from the csv file) and browse to it
    search = link
    browser.get(search)
    time.sleep(5)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    #browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    #raw = urlopen(link).read()
    #page = BeautifulSoup(raw, "html.parser")
    page = BeautifulSoup(browser.page_source, 'lxml')
    
    tcounter += 1
    
    if tcounter % 100 == 0:
        time.sleep(1000)
    
    
#Experience Section
    tall = page.find_all('li', class_ = "pv-entity__position-group-pager pv-profile-section__list-item ember-view")
    titles = []
    companies = []
    dates = []
    
    for li in tall:
        _title = li.find_all('h3','t-16 t-black t-bold')
        _companies = li.find_all('p','pv-entity__secondary-title t-14 t-black t-normal')
        _dates = li.find_all('span', class_ = "pv-entity__bullet-item-v2")
        if _title:
            titles.append(_title[0].get_text().strip()) 
        else:
             titles.append('')
        if _companies:
            companies.append(_companies[0].get_text().strip()) 
        else:
             companies.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
        print(companies)
    #Put scraped data into exp_df


    profile = link
    exp_titles = titles
    exp_companies =  companies
    exp_dates = dates
    
    #Education Section
    tall = page.find_all('li', class_ = "pv-profile-section__list-item pv-education-entity pv-profile-section__card-item ember-view")
    study_name = [] 
    degree = []
    dates = []
    
    for li in tall:
        _std_name = li.find_all('p','pv-entity__secondary-title pv-entity__fos t-14 t-black t-normal')
        _deg = li.find_all('p','pv-entity__secondary-title pv-entity__degree-name t-14 t-black t-normal')
        _dates = li.find_all('p', class_ = "pv-entity__dates t-14 t-black--light t-normal  time")
        if _std_name:
            _study_name = _std_name[0].find_all('span', 'pv-entity__comma-item')
            study_name.append(_study_name[0].get_text().strip()) 
        else:
             study_name.append('')
        if _deg:
            _degree = _deg[0].find_all('span', 'pv-entity__comma-item')
            degree.append(_degree[0].get_text().strip()) 
        else:
             degree.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
                
    ed_name = study_name
    ed_deg = degree
    ed_dates = dates
                
    #Skill Section 
    tall = page.find_all('li', class_ = "pv-skill-category-entity__top-skill pv-skill-category-entity pb3 pt4 pv-skill-endorsedSkill-entity relative ember-view")
    skill = []
    
    for li in tall:
        _skill = li.find_all('span','pv-skill-category-entity__name-text t-16 t-black t-bold')
        if _skill:
            skill.append(_skill[0].get_text().strip()) 
        else:
             skill.append('')

        
    temp1 = pd.DataFrame({'profile':profile, 'exp_title':exp_titles, 'exp_company':exp_companies, 'exp_dates':exp_dates})
    temp2 = pd.DataFrame({'profile':profile, 'ed_name':ed_name, 'ed_deg':ed_deg, 'ed_dates':ed_dates}) 
    temp3 = pd.DataFrame({'profile':profile, 'skill':skill})
    newExp_df = newExp_df.append(temp1)
    newEdu_df = newEdu_df.append(temp2)
    newSki_df = newSki_df.append(temp3)
    print(link, 'completed')
   

newExp_df.to_csv('experience2.csv')
newEdu_df.to_csv('education2.csv')
newSki_df.to_csv('skill2.csv')

### for the next time i was able to get informations from 256 others profiles.

In [ ]:
new_R = newR.tail(310)

In [ ]:
#create empty data frame
newExp_df = pd.DataFrame(columns = ['profile', 'exp_title', 'exp_company', 'exp_dates'])
newEdu_df = pd.DataFrame(columns = ['profile', 'ed_name', 'ed_deg', 'ed_dates'])
newSki_df = pd.DataFrame(columns = ['profile', 'skill'])

tcounter = 0 
#Create big loop
#for link in r.loc[0:5,'profile']:
for link in new_R.loc[:,'profile']:
    if link == 'https://www.linkedin.com#':     #if it equal link then skip
        continue 
    time.sleep(5)
# This section is where you put in the profile link (loaded from the csv file) and browse to it
    search = link
    browser.get(search)
    time.sleep(5)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    #browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    #raw = urlopen(link).read()
    #page = BeautifulSoup(raw, "html.parser")
    page = BeautifulSoup(browser.page_source, 'lxml')
    
    tcounter += 1
    
    if tcounter % 100 == 0:
        time.sleep(900)
    
    
#Experience Section
    tall = page.find_all('li', class_ = "pv-entity__position-group-pager pv-profile-section__list-item ember-view")
    titles = []
    companies = []
    dates = []
    
    for li in tall:
        _title = li.find_all('h3','t-16 t-black t-bold')
        _companies = li.find_all('p','pv-entity__secondary-title t-14 t-black t-normal')
        _dates = li.find_all('span', class_ = "pv-entity__bullet-item-v2")
        if _title:
            titles.append(_title[0].get_text().strip()) 
        else:
             titles.append('')
        if _companies:
            companies.append(_companies[0].get_text().strip()) 
        else:
             companies.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
        print(companies)
    #Put scraped data into exp_df


    profile = link
    exp_titles = titles
    exp_companies =  companies
    exp_dates = dates
    
    #Education Section
    tall = page.find_all('li', class_ = "pv-profile-section__list-item pv-education-entity pv-profile-section__card-item ember-view")
    study_name = [] 
    degree = []
    dates = []
    
    for li in tall:
        _std_name = li.find_all('p','pv-entity__secondary-title pv-entity__fos t-14 t-black t-normal')
        _deg = li.find_all('p','pv-entity__secondary-title pv-entity__degree-name t-14 t-black t-normal')
        _dates = li.find_all('p', class_ = "pv-entity__dates t-14 t-black--light t-normal  time")
        if _std_name:
            _study_name = _std_name[0].find_all('span', 'pv-entity__comma-item')
            study_name.append(_study_name[0].get_text().strip()) 
        else:
             study_name.append('')
        if _deg:
            _degree = _deg[0].find_all('span', 'pv-entity__comma-item')
            degree.append(_degree[0].get_text().strip()) 
        else:
             degree.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
                
    ed_name = study_name
    ed_deg = degree
    ed_dates = dates
                
    #Skill Section 
    tall = page.find_all('li', class_ = "pv-skill-category-entity__top-skill pv-skill-category-entity pb3 pt4 pv-skill-endorsedSkill-entity relative ember-view")
    skill = []
    
    for li in tall:
        _skill = li.find_all('span','pv-skill-category-entity__name-text t-16 t-black t-bold')
        if _skill:
            skill.append(_skill[0].get_text().strip()) 
        else:
             skill.append('')

        
    temp1 = pd.DataFrame({'profile':profile, 'exp_title':exp_titles, 'exp_company':exp_companies, 'exp_dates':exp_dates})
    temp2 = pd.DataFrame({'profile':profile, 'ed_name':ed_name, 'ed_deg':ed_deg, 'ed_dates':ed_dates}) 
    temp3 = pd.DataFrame({'profile':profile, 'skill':skill})
    newExp_df = newExp_df.append(temp1)
    newEdu_df = newEdu_df.append(temp2)
    newSki_df = newSki_df.append(temp3)
    print(link, 'completed')
    print(tcounter)
    if tcounter  == 210:
        break
   

newExp_df.to_csv('experience3.csv')
newEdu_df.to_csv('education3.csv')
newSki_df.to_csv('skill3.csv')

### In that time i was able to get informations from 200 profiles.

In [ ]:
done_R = new_R.tail(110)

In [ ]:
#create empty data frame
newExp_df = pd.DataFrame(columns = ['profile', 'exp_title', 'exp_company', 'exp_dates'])
newEdu_df = pd.DataFrame(columns = ['profile', 'ed_name', 'ed_deg', 'ed_dates'])
newSki_df = pd.DataFrame(columns = ['profile', 'skill'])

tcounter = 0 
#Create big loop
#for link in r.loc[0:5,'profile']:
for link in done_R.loc[:,'profile']:
    if link == 'https://www.linkedin.com#':     #if it equal link then skip
        continue 
    time.sleep(5)
# This section is where you put in the profile link (loaded from the csv file) and browse to it
    search = link
    browser.get(search)
    time.sleep(5)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.PAGE_DOWN)
    #browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    #raw = urlopen(link).read()
    #page = BeautifulSoup(raw, "html.parser")
    page = BeautifulSoup(browser.page_source, 'lxml')
    
    tcounter += 1
    
    if tcounter % 100 == 0:
        time.sleep(900)
    
    
#Experience Section
    tall = page.find_all('li', class_ = "pv-entity__position-group-pager pv-profile-section__list-item ember-view")
    titles = []
    companies = []
    dates = []
    
    for li in tall:
        _title = li.find_all('h3','t-16 t-black t-bold')
        _companies = li.find_all('p','pv-entity__secondary-title t-14 t-black t-normal')
        _dates = li.find_all('span', class_ = "pv-entity__bullet-item-v2")
        if _title:
            titles.append(_title[0].get_text().strip()) 
        else:
             titles.append('')
        if _companies:
            companies.append(_companies[0].get_text().strip()) 
        else:
             companies.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
    #Put scraped data into exp_df


    profile = link
    exp_titles = titles
    exp_companies =  companies
    exp_dates = dates
    
    #Education Section
    tall = page.find_all('li', class_ = "pv-profile-section__list-item pv-education-entity pv-profile-section__card-item ember-view")
    study_name = [] 
    degree = []
    dates = []
    
    for li in tall:
        _std_name = li.find_all('p','pv-entity__secondary-title pv-entity__fos t-14 t-black t-normal')
        _deg = li.find_all('p','pv-entity__secondary-title pv-entity__degree-name t-14 t-black t-normal')
        _dates = li.find_all('p', class_ = "pv-entity__dates t-14 t-black--light t-normal  time")
        if _std_name:
            _study_name = _std_name[0].find_all('span', 'pv-entity__comma-item')
            study_name.append(_study_name[0].get_text().strip()) 
        else:
             study_name.append('')
        if _deg:
            _degree = _deg[0].find_all('span', 'pv-entity__comma-item')
            degree.append(_degree[0].get_text().strip()) 
        else:
             degree.append('')
        if _dates:
            dates.append(_dates[0].get_text().strip()) 
        else:
             dates.append('')
                
    ed_name = study_name
    ed_deg = degree
    ed_dates = dates
                
    #Skill Section 
    tall = page.find_all('li', class_ = "pv-skill-category-entity__top-skill pv-skill-category-entity pb3 pt4 pv-skill-endorsedSkill-entity relative ember-view")
    skill = []
    
    for li in tall:
        _skill = li.find_all('span','pv-skill-category-entity__name-text t-16 t-black t-bold')
        if _skill:
            skill.append(_skill[0].get_text().strip()) 
        else:
             skill.append('')

        
    temp1 = pd.DataFrame({'profile':profile, 'exp_title':exp_titles, 'exp_company':exp_companies, 'exp_dates':exp_dates})
    temp2 = pd.DataFrame({'profile':profile, 'ed_name':ed_name, 'ed_deg':ed_deg, 'ed_dates':ed_dates}) 
    temp3 = pd.DataFrame({'profile':profile, 'skill':skill})
    newExp_df = newExp_df.append(temp1)
    newEdu_df = newEdu_df.append(temp2)
    newSki_df = newSki_df.append(temp3)
    print(link, 'completed')
    print(tcounter)
   

newExp_df.to_csv('experience4.csv')
newEdu_df.to_csv('education4.csv')
newSki_df.to_csv('skill4.csv')

In [ ]:
 browser.quit()

### Finally i was able to get informations from all the profiles, of course those that had informations. Then i have 12 datasets, 4 for each, experience, education and skill.

## step 3

load these datasets in 3 datasets, for each category.

In [6]:
csv_file_list = ["experience.csv", "experience2.csv", "experience3.csv", "experience4.csv"]

list_of_dataframes = []
for filename in csv_file_list:
    list_of_dataframes.append(pd.read_csv(filename))

experience_df = pd.concat(list_of_dataframes)

csv_file_list = ["education.csv", "education2.csv", "education3.csv", "education4.csv"]

list_of_dataframes = []
for filename in csv_file_list:
    list_of_dataframes.append(pd.read_csv(filename))

education_df = pd.concat(list_of_dataframes)

csv_file_list = ["skill.csv", "skill2.csv", "skill3.csv", "skill4.csv"]

list_of_dataframes = []
for filename in csv_file_list:
    list_of_dataframes.append(pd.read_csv(filename))

skill_df = pd.concat(list_of_dataframes)

In [11]:
experience_df.drop('Unnamed: 0', axis = 1, inplace = True)
education_df.drop('Unnamed: 0', axis = 1, inplace = True)
skill_df.drop('Unnamed: 0', axis = 1, inplace = True)

In [12]:
experience_df

profile  \
0    https://www.linkedin.com/in/joseph-melior-3a4a...   
1    https://www.linkedin.com/in/alexandre-michel-0...   
2    https://www.linkedin.com/in/alexandre-michel-0...   
3    https://www.linkedin.com/in/biento-jacques-871...   
4    https://www.linkedin.com/in/valery-fils-aime-6...   
..                                                 ...   
381  https://www.linkedin.com/in/markenley-belly-sa...   
382  https://www.linkedin.com/in/markenley-belly-sa...   
383  https://www.linkedin.com/in/markenley-belly-sa...   
384  https://www.linkedin.com/in/christelle-daniel-...   
385  https://www.linkedin.com/in/christelle-daniel-...   

                                             exp_title  \
0                                                  CEO   
1                                            Rédacteur   
2                                   Directeur exécutif   
3                                           Co-Founder   
4                Co-fondateur et Resp. de partenariats   
..                                                 ...   
381                                Procurement Officer   
382                                Procurement Officer   
383                                        Logistigien   
384                Marketing Communication Coordinator   
385  Mobile Advertising Account Executive / AD Agen...   

                           exp_company     exp_dates  
0                       CHOCOMAX-HAITI   6 yrs 7 mos  
1               Journal l'Union, Haïti   2 yrs 9 mos  
2    Tras-Haïti\n        Self-employed   2 yrs 5 mos  
3                            GBI Haiti    1 yr 9 mos  
4      Haiti Climat\n        Part-time  2 yrs 10 mos  
..                                 ...           ...  
381                 American Red Cross         9 yrs  
382                       World Vision        10 mos  
383                              Oxfam   2 yrs 5 mos  
384      Alternative Insurance Company   3 yrs 7 mos  
385                      Digicel Group   2 yrs 9 mos  

[3210 rows x 4 columns]

In [13]:
education_df

profile  \
0    https://www.linkedin.com/in/joseph-melior-3a4a...   
1    https://www.linkedin.com/in/joseph-melior-3a4a...   
2    https://www.linkedin.com/in/alexandre-michel-0...   
3    https://www.linkedin.com/in/biento-jacques-871...   
4    https://www.linkedin.com/in/biento-jacques-871...   
..                                                 ...   
199              https://www.linkedin.com/in/dexavier/   
200              https://www.linkedin.com/in/dexavier/   
201  https://www.linkedin.com/in/jhon-kerby-gerlin-...   
202  https://www.linkedin.com/in/ernest-s-sherman-2...   
203  https://www.linkedin.com/in/markenley-belly-sa...   

                                               ed_name  \
0                                                  Law   
1                                           Sociologie   
2    Sciences de la Communication et des relations ...   
3                                          Photography   
4                                 Broadcast Journalism   
..                                                 ...   
199                                       INFORMATIQUE   
200    Network and System Administration/Administrator   
201    Business Administration and Management, General   
202  Logistics, Materials, and Supply Chain Management   
203                                                NaN   

                                    ed_deg  ed_dates  
0                        Bachelor's degree       NaN  
1                        Licentiate degree       NaN  
2                                  Licence       NaN  
3                                  Diplome       NaN  
4                        Bachelor's degree       NaN  
..                                     ...       ...  
199        GESTION DE SYSTEME INFORMATISEE       NaN  
200                    Computer Networking       NaN  
201                                    2nd       NaN  
202  Certificate in Humanitarian Logistics       NaN  
203                      Licentiate degree       NaN  

[2003 rows x 4 columns]

In [14]:
skill_df

profile               skill
0    https://www.linkedin.com/in/joseph-melior-3a4a...  Strategic Planning
1    https://www.linkedin.com/in/joseph-melior-3a4a...     Public Speaking
2    https://www.linkedin.com/in/joseph-melior-3a4a...   Business Strategy
3    https://www.linkedin.com/in/biento-jacques-871...      Graphic Design
4    https://www.linkedin.com/in/biento-jacques-871...    Video Production
..                                                 ...                 ...
246  https://www.linkedin.com/in/markenley-belly-sa...          Government
247  https://www.linkedin.com/in/markenley-belly-sa...          Nonprofits
248  https://www.linkedin.com/in/christelle-daniel-...          Management
249  https://www.linkedin.com/in/christelle-daniel-...        Direct Sales
250  https://www.linkedin.com/in/christelle-daniel-...      Salesforce.com

[2258 rows x 2 columns]